### PR TITLE
block3 dop queue

### PR DESCRIPTION
**Pull request by DenisBilobram**

### CODE REVIEW

File: queue.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
+#include <iostream>
+#include <stack>
+#include <algorithm>
+#include <functional>
+
+using namespace std;
```
Comment (Hyperb0rean) [inline queue.cpp]: Cringe

File: queue.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+template<typename T, typename F = function<T(T, T)>>
```
Comment (Hyperb0rean) [inline queue.cpp]: Default template parameter for F is useless

File: queue.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+class AggQueue {
+private:
+    stack<pair<T, T>> inputStack, outputStack;
```
Comment (Hyperb0rean) [inline queue.cpp]: Coupling of actual and dynamic values is not a good thing tbh, as i described to other student:
E.g. you want `emplace(Args&&)` functuon, you cant do this one with coupled stacks.

File: queue.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    void pushWithAgg(stack<pair<T, T>>& stack, T value) {
+        T currentAgg = stack.empty() ? value : func(value, stack.top().second);
+        stack.push({value, currentAgg});
+    }
+
+    void transferIfNeeded() {
+        if (outputStack.empty()) {
+            while (!inputStack.empty()) {
+                T element = inputStack.top().first;
+                pushWithAgg(outputStack, element);
```
Comment (Hyperb0rean) [inline queue.cpp]: Is it needed to call functor twice?

File: queue.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+public:
+    AggQueue(F f) : func(f) {}
+
+    void enqueue(T value) {
+        pushWithAgg(inputStack, value);
+    }
+
+    void dequeue() {
+        if (inputStack.empty() && outputStack.empty()) {
+            cout << "Queue is empty! Cannot dequeue." << endl;
```
Comment (Hyperb0rean) [inline queue.cpp]: Lets throw (C++ way), or lets return  `std::optional<T>` (Personally I like Result/Optional error handling approach more, but up to you)

File: queue.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
+#include <iostream>
+#include <stack>
+#include <algorithm>
+#include <functional>
+
+using namespace std;
+
+template<typename T, typename F = function<T(T, T)>>
+class AggQueue {
```
Comment (Hyperb0rean) [inline queue.cpp]: Lets do not shorten names `Agg -> Aggregate`

File: queue.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            cout << "Queue is empty! Cannot dequeue." << endl;
+            return;
+        }
+        transferIfNeeded();
+        outputStack.pop();
+    }
+
+    T getAgg() {
+        if (inputStack.empty() && outputStack.empty()) {
+            cout << "Queue is empty!" << endl;
```
Comment (Hyperb0rean) [inline queue.cpp]: Not needed, as above

File: queue.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        if (inputStack.empty() && outputStack.empty()) {
+            cout << "Queue is empty! Cannot dequeue." << endl;
+            return;
+        }
+        transferIfNeeded();
+        outputStack.pop();
+    }
+
+    T getAgg() {
+        if (inputStack.empty() && outputStack.empty()) {
```
Comment (Hyperb0rean) [inline queue.cpp]: Also use your `isEmpty()`

File: queue.cpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    bool isEmpty() {
+        return inputStack.empty() && outputStack.empty();
+    }
+};
+
+int main() {
+
+    function<int(int, int)> minFunc =  [](int a, int b) { return min(a, b); };
```
Comment (Hyperb0rean) [inline queue.cpp]: Usong `auto` with lambdas is idiomatic way

File: queue.cpp [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    function<int(int, int)> maxFunc =  [](int a, int b) { return max(a, b); };
+    function<int(int, int)> sumFunc =  [](int a, int b) { return a + b; };
+
+    AggQueue<int, function<int(int, int)>> minQueue(minFunc);
```
Comment (Hyperb0rean) [inline queue.cpp]: Template parameter for Functor will be deduced

File: queue.cpp [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    AggQueue<int, function<int(int, int)>> maxQueue(maxFunc);
+    AggQueue<int, function<int(int, int)>> sumQueue(sumFunc);
+
+    minQueue.enqueue(5);
+    minQueue.enqueue(3);
+    minQueue.enqueue(1);
+    minQueue.enqueue(7);
+    cout << "Minimum: " << minQueue.getAgg() << endl; // Ожидается 1
```
Comment (Hyperb0rean) [inline queue.cpp]: use asserts, or even better write separate testing function

**Review by Hyperb0rean — CHANGES_REQUESTED** [review]
I have comments

File: queue.cpp (hunk 4 above)
Comment (DenisBilobram) [reply queue.cpp]: Yes, because inputStack and outputStack will store different aggregate values due to the different order of adding values.

**Comment by DenisBilobram** [comment]
@Hyperb0rean fixed the program according to your comments

File: queue.cpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        stack.push(value);
+        aggStack.push(currentAgg);
+    }
+
+    void transferIfNeeded() {
+        if (outputStack.empty()) {
+            while (!inputStack.empty()) {
+                T element = inputStack.top();
+                inputStack.pop();
+                pushWithAgg(outputStack, outputAggStack, element);
```
Comment (Hyperb0rean) [inline queue.cpp:26]: The same is it was before, you can just pass aggregated value to second stack without calling func again every time

File: queue.cpp [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+                inputAggStack.pop();
+            }
+        }
+    }
+
+public:
+    AggregateQueue(F f) : func(f) {}
```
Comment (Hyperb0rean) [inline queue.cpp]: Add `explicit`

File: queue.cpp [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        if (inputStack.empty()) {
+            return outputAggStack.top();
+        }
+        if (outputStack.empty()) {
+            return inputAggStack.top();
+        }
+        return func(outputAggStack.top(), inputAggStack.top());
+    }
+
+    T peek() {
```
Comment (Hyperb0rean) [inline queue.cpp]: Do back and front

File: queue.cpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        return inputStack.empty() && outputStack.empty();
+    }
+};
+
+int main() {
+    auto minFunc = [](int a, int b) { return std::min(a, b); };
+    auto maxFunc = [](int a, int b) { return std::max(a, b); };
+    auto sumFunc = [](int a, int b) { return a + b; };
+
+    AggregateQueue<int, decltype(minFunc)> minQueue(minFunc);
```
Comment (Hyperb0rean) [inline queue.cpp]: Just pass lambda to the constructor, type should be deduced, i think

**Review by Hyperb0rean — CHANGES_REQUESTED** [review]
I have comments

File: queue.cpp (hunk 12 above)
Comment (DenisBilobram) [reply queue.cpp:26]: I can't understand your point. As I mentioned earlier, when transferring values from the inputStack to the outputStack, their order is reversed, which makes it necessary to recalculate the aggregated value in the new order. This is clearly noticeable in the example of a queue with an aggregated sum of elements.

**Comment by DenisBilobram** [comment]
@Hyperb0rean I have fixed program according to your comments, except for the one (comment) I responded to.

File: queue.cpp (hunk 12 above)
Comment (Hyperb0rean) [reply queue.cpp:26]: Ah, ok I can see, btw there is possibility to make it better, Ill tell you during the class

**Review by Hyperb0rean — APPROVED** [review]
LGTM

**Pull request merged**